// reflex/dist/reflex.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "./task", "./effects", "./dom", "./subscription"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("./task"), require("./effects"), require("./dom"), require("./subscription"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.task, global.effects, global.dom, global.subscription);
    global.application = mod.exports;
  }
})(this, function (exports, _task, _effects, _dom, _subscription) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.start = exports.beginner = undefined;


  class Application {
    constructor(send, model, view, task, services) {
      this.send = send;
      this.model = model;
      this.view = view;
      this.task = task;
      this.services = services;
    }

  }

  const first = xs => xs[0];
  const second = xs => xs[1];

  const beginner = exports.beginner = configuration => ({
    flags: void 0,
    init: _ => [configuration.model, _effects.Effects.none],
    update: (model, action) => [configuration.update(model, action), _effects.Effects.none],
    view: configuration.view,
    subscriptions: model => _subscription.none
  });

  const noSubscribtions = _ => _subscription.none;

  const start = exports.start = configuration => write => {
    const { init, view, update, flags } = configuration;
    const subscriptions = configuration.subscriptions == null ? noSubscribtions : configuration.subscriptions;

    const send = action => {
      const [model, fx] = update(application.model, action);
      application.model = model;
      application.view = (0, _dom.root)(view, model, send);
      application.task = fx.execute(send);

      application.services = subscriptions(model).reduce(subscribe, application.services);
      exectueServices(application.services, send);
      write(application);
    };

    const [model, fx] = init(flags);

    const application = new Application(send, model, (0, _dom.root)(view, model, send), fx.execute(send), subscriptions(model).reduce(subscribe, {
      nextAddress: 0,
      active: Object.create(null)
    }));

    exectueServices(application.services, send);
    return write(application);
  };

  const subscribe = (services, subscription) => {
    const { active } = services;
    const { feed, detail } = subscription;
    const service = active[String(feed.address)];

    if (service == null || service.feed !== feed) {
      const address = `/${++services.nextAddress}`;
      const subscriptions = [subscription];
      const state = feed.init();
      feed.address = address;

      active[address] = { feed, subscriptions, state };
    } else {
      service.subscriptions.push(subscription);
    }

    return services;
  };

  const exectueServices = (services, send) => {
    for (let address in services.active) {
      exectueService(services.active[address], send);
    }
  };

  const exectueService = (service, send) => {
    const { state, feed, subscriptions } = service;
    const [model, fx] = feed.update(state, feed.subscribe(subscriptions));
    service.state = model;
    fx.execute(send);
  };
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "./driver"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("./driver"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.driver);
    global.dom = mod.exports;
  }
})(this, function (exports, _driver) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.thunk = exports.node = exports.text = exports.root = exports.LazyNode = undefined;


  let driver = null;

  const absent = new String("absent");

  class Root {
    constructor(view, model, address) {
      _initialiseProps.call(this);

      this.view = view;
      this.model = model;
      this.address = address;
    }
    renderWith(current) {
      let exception = absent;
      const previous = driver;
      driver = current;

      try {
        driver.render(this.view(this.model, this.address));
      } catch (error) {
        exception = error;
      }

      driver = previous;

      if (exception != absent) {
        throw exception;
      }
    }

  }

  var _initialiseProps = function () {
    this.$type = "VirtualRoot";
  };

  class LazyNode {
    constructor(tagName, properties, children) {
      this.$type = "LazyTree";

      this.tagName = tagName;
      this.properties = properties;
      this.children = children;
      this.key = properties == null ? null : properties.key;
    }
    force() {
      if (driver == null) {
        throw TypeError("LazyTree may only be forced from with in the Root.renderWith(driver) call");
      }

      return driver.node(this.tagName, this.properties, this.children);
    }

  }

  exports.LazyNode = LazyNode;
  class LazyThunk {
    constructor(key, view, ...args) {
      this.$type = "LazyTree";

      this.key = key;
      this.view = view;
      this.args = args;
    }
    force() {
      if (driver == null) {
        throw TypeError("LazyTree may only be forced from with in the Root.renderWith(driver) call");
      }

      return driver.thunk(this.key, this.view, ...this.args);
    }

  }

  const root = exports.root = (view, model, address) => new Root(view, model, address);

  const text = exports.text = content => driver == null ? content : driver.text(content);

  const node = exports.node = (tagName, properties, children) => driver == null ? new LazyNode(tagName, properties, children) : driver.node(tagName, properties, children);

  const thunk = exports.thunk = (key, view, a0, a1, a2, a3, a4, a5, a6, a7, a8, a9) => driver == null ? new LazyThunk(key, view, a0, a1, a2, a3, a4, a5, a6, a7, a8, a9) : driver.thunk(key, view, a0, a1, a2, a3, a4, a5, a6, a7, a8, a9);
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define([], factory);
  } else if (typeof exports !== "undefined") {
    factory();
  } else {
    var mod = {
      exports: {}
    };
    factory();
    global.driver = mod.exports;
  }
})(this, function () {
  "use strict";
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "./task"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("./task"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.task);
    global.effects = mod.exports;
  }
})(this, function (exports, _task) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.Effects = undefined;


  const raise = error => {
    throw Error(`Effects should be created from task that never fail but it did fail with error ${error}`);
  };

  const ignore = _ => void 0;

  const nil = _task.Task.succeed(void 0);

  const never = new _task.Task((succeed, fail) => void 0);

  class Effects {
    static task(task) {
      console.warn("Effects.task is deprecated please use Effects.perform instead");
      return new Perform(task);
    }
    static perform(task) {
      return new Perform(task);
    }
    static tick(tag) {
      console.warn("Effects.tick is deprecated please use Effects.perform(Task.requestAnimationFrame().map(tag)) instead");
      return new Perform(_task.Task.requestAnimationFrame().map(tag));
    }
    static receive(action) {
      const fx = new Perform(new _task.Task((succeed, fail) => void Promise.resolve(action).then(succeed, fail)));
      return fx;
    }
    static batch(effects) {
      return new Batch(effects);
    }
    map(f) {
      throw Error("Subclass of abstract Effect must implement map");
    }
    execute(address) {
      throw Error("Subclass of abstract Effect must implement execute");
    }
  }

  exports.Effects = Effects;
  class Perform extends Effects {
    constructor(task) {
      super();
      this.task = task;
    }
    map(f) {
      return new Perform(this.task.map(f));
    }
    execute(address) {
      return this.task.chain(value => _task.Task.send(address, value));
    }
  }

  class None extends Effects {
    map(f) {
      return Effects.none;
    }
    execute(address) {
      return nil;
    }
  }
  Effects.none = new None();

  class Batch extends Effects {
    constructor(effects) {
      super();
      this.effects = effects;
    }
    map(f) {
      return new Batch(this.effects.map(effect => effect.map(f)));
    }
    execute(address) {
      return new _task.Task((succeed, fail) => {
        const { effects } = this;
        const count = effects.length;
        let index = 0;
        while (index < count) {
          const effect = effects[index];
          if (!(effect instanceof None)) {
            _task.Task.fork(effect.execute(address), ignore, raise);
          }

          index = index + 1;
        }
        succeed(void 0);
      });
    }

  }
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "./dom"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("./dom"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.dom);
    global.html = mod.exports;
  }
})(this, function (exports, _dom) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.html = undefined;
  const html = exports.html = ["a", "abbr", "address", "area", "article", "aside", "audio", "b", "base", "bdi", "bdo", "big", "blockquote", "body", "br", "button", "canvas", "caption", "cite", "code", "col", "colgroup", "data", "datalist", "dd", "del", "details", "dfn", "dialog", "div", "dl", "dt", "em", "embed", "fieldset", "figcaption", "figure", "footer", "form", "h1", "h2", "h3", "h4", "h5", "h6", "head", "header", "hr", "html", "i", "iframe", "img", "input", "ins", "kbd", "keygen", "label", "legend", "li", "link", "main", "map", "mark", "menu", "menuitem", "meta", "meter", "nav", "noscript", "object", "ol", "optgroup", "option", "output", "p", "param", "picture", "pre", "progress", "q", "rp", "rt", "ruby", "s", "samp", "script", "section", "select", "small", "source", "span", "strong", "style", "sub", "summary", "sup", "table", "tbody", "td", "textarea", "tfoot", "th", "thead", "time", "title", "tr", "track", "u", "ul", "var", "video", "wbr", "circle", "clipPath", "defs", "ellipse", "g", "line", "linearGradient", "mask", "path", "pattern", "polygon", "polyline", "radialGradient", "rect", "stop", "svg", "text", "tspan"].reduce((html, tagName) => {
    const element = (properties, children) => (0, _dom.node)(tagName, properties, children);

    html[tagName] = element;
    return html;
  }, (
  // @Hack: Workarounding facebook/flow#1683
  Object.create(null), {}));
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "./driver", "./signal", "./subscription", "./dom", "./html", "./application", "./task", "./effects"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("./driver"), require("./signal"), require("./subscription"), require("./dom"), require("./html"), require("./application"), require("./task"), require("./effects"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.driver, global.signal, global.subscription, global.dom, global.html, global.application, global.task, global.effects);
    global.index = mod.exports;
  }
})(this, function (exports, _driver, _signal, _subscription, _dom, _html, _application, _task, _effects) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.Subscription = exports.Driver = exports.Effects = exports.Task = exports.beginner = exports.start = exports.html = exports.root = exports.thunk = exports.text = exports.node = exports.subscription = exports.forward = exports.VirtualRoot = exports.Address = exports.VirtualTree = exports.LazyTree = exports.Widget = exports.Thunk = exports.VirtualNode = exports.VirtualText = exports.StyleDictionary = exports.AttributeDictionary = exports.PropertyDictionary = undefined;
  Object.defineProperty(exports, "PropertyDictionary", {
    enumerable: true,
    get: function () {
      return _driver.PropertyDictionary;
    }
  });
  Object.defineProperty(exports, "AttributeDictionary", {
    enumerable: true,
    get: function () {
      return _driver.AttributeDictionary;
    }
  });
  Object.defineProperty(exports, "StyleDictionary", {
    enumerable: true,
    get: function () {
      return _driver.StyleDictionary;
    }
  });
  Object.defineProperty(exports, "VirtualText", {
    enumerable: true,
    get: function () {
      return _driver.VirtualText;
    }
  });
  Object.defineProperty(exports, "VirtualNode", {
    enumerable: true,
    get: function () {
      return _driver.VirtualNode;
    }
  });
  Object.defineProperty(exports, "Thunk", {
    enumerable: true,
    get: function () {
      return _driver.Thunk;
    }
  });
  Object.defineProperty(exports, "Widget", {
    enumerable: true,
    get: function () {
      return _driver.Widget;
    }
  });
  Object.defineProperty(exports, "LazyTree", {
    enumerable: true,
    get: function () {
      return _driver.LazyTree;
    }
  });
  Object.defineProperty(exports, "VirtualTree", {
    enumerable: true,
    get: function () {
      return _driver.VirtualTree;
    }
  });
  Object.defineProperty(exports, "Address", {
    enumerable: true,
    get: function () {
      return _driver.Address;
    }
  });
  Object.defineProperty(exports, "VirtualRoot", {
    enumerable: true,
    get: function () {
      return _driver.VirtualRoot;
    }
  });
  Object.defineProperty(exports, "forward", {
    enumerable: true,
    get: function () {
      return _signal.forward;
    }
  });
  Object.defineProperty(exports, "subscription", {
    enumerable: true,
    get: function () {
      return _subscription.subscription;
    }
  });
  Object.defineProperty(exports, "node", {
    enumerable: true,
    get: function () {
      return _dom.node;
    }
  });
  Object.defineProperty(exports, "text", {
    enumerable: true,
    get: function () {
      return _dom.text;
    }
  });
  Object.defineProperty(exports, "thunk", {
    enumerable: true,
    get: function () {
      return _dom.thunk;
    }
  });
  Object.defineProperty(exports, "root", {
    enumerable: true,
    get: function () {
      return _dom.root;
    }
  });
  Object.defineProperty(exports, "html", {
    enumerable: true,
    get: function () {
      return _html.html;
    }
  });
  Object.defineProperty(exports, "start", {
    enumerable: true,
    get: function () {
      return _application.start;
    }
  });
  Object.defineProperty(exports, "beginner", {
    enumerable: true,
    get: function () {
      return _application.beginner;
    }
  });
  Object.defineProperty(exports, "Task", {
    enumerable: true,
    get: function () {
      return _task.Task;
    }
  });
  Object.defineProperty(exports, "Effects", {
    enumerable: true,
    get: function () {
      return _effects.Effects;
    }
  });

  var Driver = _interopRequireWildcard(_driver);

  var Subscription = _interopRequireWildcard(_subscription);

  function _interopRequireWildcard(obj) {
    if (obj && obj.__esModule) {
      return obj;
    } else {
      var newObj = {};

      if (obj != null) {
        for (var key in obj) {
          if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key];
        }
      }

      newObj.default = obj;
      return newObj;
    }
  }

  exports.Driver = Driver;
  exports.Subscription = Subscription;
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports);
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports);
    global.preemptiveAnimationFrame = mod.exports;
  }
})(this, function (exports) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });


  // Invariants:
  // 1. In the NO_REQUEST state, there is never a scheduled animation frame.
  // 2. In the PENDING_REQUEST and EXTRA_REQUEST states, there is always exactly
  // one scheduled animation frame.
  const NO_REQUEST = 0;
  const PENDING_REQUEST = 1;
  const EXTRA_REQUEST = 2;

  let nextID = 0;
  let state = NO_REQUEST;
  let requests = [];
  let ids = [];

  const absent = new String("absent");

  const requestAnimationFrame = exports.requestAnimationFrame = request => {
    if (state === NO_REQUEST) {
      window.requestAnimationFrame(performAnimationFrame);
    }

    const id = ++nextID;
    requests.push(request);
    ids.push(id);
    state = PENDING_REQUEST;
    return id;
  };

  const cancelAnimationFrame = exports.cancelAnimationFrame = id => {
    const index = ids.indexOf(id);
    if (index >= 0) {
      ids.splice(index, 1);
      requests.splice(index, 1);
    }
  };

  const forceAnimationFrame = exports.forceAnimationFrame = (time = window.performance.now()) => performAnimationFrame(time);

  const performAnimationFrame = time => {
    switch (state) {
      case NO_REQUEST:
        // This state should not be possible. How can there be no
        // request, yet somehow we are actively fulfilling a
        // request?
        throw Error(`Unexpected frame request`);
      case PENDING_REQUEST:
        // At this point, we do not *know* that another frame is
        // needed, but we make an extra frame request just in
        // case. It's possible to drop a frame if frame is requested
        // too late, so we just do it preemptively.
        window.requestAnimationFrame(performAnimationFrame);
        state = EXTRA_REQUEST;
        ids.splice(0);
        dispatchAnimationFrame(requests.splice(0), 0, time);
        break;
      case EXTRA_REQUEST:
        // Turns out the extra request was not needed, so we will
        // stop requesting. No reason to call it all the time if
        // no one needs it.
        state = NO_REQUEST;
        break;
    }
  };

  const dispatchAnimationFrame = (requests, index, time) => {
    let exception = absent;
    const count = requests.length;
    try {
      while (index < count) {
        const request = requests[index];
        index = index + 1;
        request(time);
      }
    } catch (error) {
      exception = error;
    }

    if (index < count) {
      dispatchAnimationFrame(requests, index, time);
    }

    if (exception != absent) {
      throw exception;
    }
  };
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports);
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports);
    global.signal = mod.exports;
  }
})(this, function (exports) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });


  const Forward = (address, tag) => {
    const forward = message => address(tag(message));
    forward.to = address;
    forward.tag = tag;
    return forward;
  };


  if (global["reflex/address"] == null) {
    global["reflex/address"] = 0;
  }

  // Create a new address. This address will tag each message it receives and then
  // forward it along to the given address.
  // Example:
  //
  // const Remove = target => {type: "Remove", target}
  // removeAddress = forward(address, Remove)
  //
  // Above example created `removeAddress` tags each message with `Remove` tag
  // before forwarding them to a general `address`.
  const forward = exports.forward = (address, tag) => {
    // Genrate ID for each address that has a forwarding addresses so that
    // forwarding addresses could be cached by that id and a tag-ing function.
    const id = address.id != null ? address.id : address.id = global["reflex/address"]++;
    const key = `reflex/address/${id}`;

    return tag[key] || (tag[key] = Forward(address, tag));
  };
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "./Effects"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("./Effects"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.Effects);
    global.subscription = mod.exports;
  }
})(this, function (exports, _Effects) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.none = exports.batch = exports.subscription = undefined;


  class Subscription {
    map(tag) {
      return this.map(tag);
    }
    reduce(reducer, init) {
      return init;
    }
  }

  class Subscribe extends Subscription {
    constructor(feed, detail, decoder) {
      super();
      this.feed = feed;
      this.detail = detail;
      this.decoder = decoder;
    }
    map(tag) {
      const decoder = value => tag(this.decoder(value));
      return new Subscribe(this.feed, this.detail, decoder);
    }
    reduce(reducer, init) {
      return reducer(init, this);
    }

  }

  class Batch extends Subscription {
    constructor(subscriptions) {
      super();
      this.subscriptions = subscriptions;
    }
    map(tag) {
      const subscriptions = this.subscriptions.map($ => $.map(tag));
      return new Batch(subscriptions);
    }
    reduce(reducer, init) {
      return this.subscriptions.reduce((result, subscription) => subscription.reduce(reducer, result), init);
    }
  }

  const subscription = exports.subscription = (feed, detail, decoder) => new Subscribe(feed, detail, decoder);

  const batch = exports.batch = subscriptions => new Batch(subscriptions);

  const none = exports.none = new Batch([]);
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "./preemptive-animation-frame"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("./preemptive-animation-frame"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.preemptiveAnimationFrame);
    global.task = mod.exports;
  }
})(this, function (exports, _preemptiveAnimationFrame) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.Task = undefined;
  const raise = error => {
    throw Error(`Task was not supposet to never fail but it did fail with error ${error}`);
  };

  const ignore = _ => void 0;

  class Task {
    static create(execute) {
      console.warn("Task.create is deprecated API use new Task instead");
      return new Task(execute);
    }
    static future(request) {
      console.warn("Task.future is deprecated API use new Task instead");
      return new Future(request);
    }
    static succeed(value) {
      return new Succeed(value);
    }

    static fail(error) {
      return new Fail(error);
    }

    static spawn(task) {
      return new Spawn(task);
    }

    static sleep(time) {
      return new Sleep(time);
    }

    static requestAnimationFrame() {
      return new AnimationFrame();
    }

    static send(address, message) {
      return new Send(address, message);
    }

    static fork(task, onSucceed, onFail) {
      return Thread.fork(task, onSucceed, onFail);
    }

    static perform(task) {
      Thread.fork(task, ignore, raise);
    }

    constructor(execute, cancel) {
      this.type = "Task";
      const task = this;
      if (execute != null) {
        task.fork = execute;
      }
      if (cancel != null) {
        task.abort = cancel;
      }
    }
    chain(next) {
      return new Chain(this, next);
    }
    map(f) {
      return new Map(this, f);
    }
    capture(handle) {
      return new Capture(this, handle);
    }
    format(f) {
      return new Format(this, f);
    }
    recover(regain) {
      return new Recover(this, regain);
    }
    fork(succeed, fail) {
      return this.execute(succeed, fail);
    }
    abort(token) {
      return this.cancel(token);
    }

  }

  exports.Task = Task;
  class Succeed extends Task {
    constructor(value) {
      super();
      this.type = "Succeed";
      this.value = value;
    }
    fork(succeed, fail) {
      succeed(this.value);
    }

  }

  class Fail extends Task {
    constructor(error) {
      super();
      this.type = "Fail";
      this.error = error;
    }
    fork(succeed, fail) {
      fail(this.error);
    }

  }

  class Sleep extends Task {
    constructor(time) {
      super();
      this.time = time;
    }
    fork(succeed, fail) {
      return setTimeout(succeed, this.time, void 0);
    }
    abort(id) {
      clearTimeout(id);
    }

  }

  class AnimationFrame extends Task {
    constructor() {
      super();
    }
    fork(succeed, fail) {
      return (0, _preemptiveAnimationFrame.requestAnimationFrame)(succeed);
    }
    abort(id) {
      (0, _preemptiveAnimationFrame.cancelAnimationFrame)(id);
    }
  }

  let threadID = 0;
  class Spawn extends Task {
    constructor(task) {
      super();
      this.task = task;
    }
    fork(succeed, fail) {
      Promise.resolve(null).then(_ => Task.fork(this.task, noop, noop));

      succeed(++threadID);
    }

  }

  class Send extends Task {
    constructor(address, message) {
      super();
      this.message = message;
      this.address = address;
    }
    fork(succeed, fail) {
      succeed(void this.address(this.message));
    }

  }

  class Future extends Task {
    constructor(request) {
      super();
      this.request = request;
    }
    fork(succeed, fail) {
      this.request().then(succeed, fail);
    }

  }

  class Then extends Task {
    constructor(task) {
      super();
      this.type = "Then";
      this.task = task;
    }
    fork(succeed, fail) {
      this.task.fork(value => void this.next(value).fork(succeed, fail), fail);
    }
    next(input) {
      throw Error("Subclass of absract Then must implement next method");
    }

  }

  class Chain extends Then {
    constructor(task, next) {
      super(task);
      this.chainer = next;
    }
    next(input) {
      return this.chainer(input);
    }

  }

  class Map extends Then {
    constructor(task, mapper) {
      // Note: Had to trick flow into thinking that `Format.prototype.handle` was
      // passed, otherwise it fails to infer polymorphic nature.
      super(task);
      this.mapper = mapper;
    }
    next(input) {
      return new Succeed(this.mapper(input));
    }

  }

  class Catch extends Task {
    constructor(task) {
      super();
      this.type = "Catch";
      this.task = task;
    }
    fork(succeed, fail) {
      this.task.fork(succeed, error => void this.handle(error).fork(succeed, fail));
    }
    handle(error) {
      throw Error("Subclass of absract Catch must implement handle method");
    }

  }

  class Capture extends Catch {
    constructor(task, handle) {
      super(task);
      this.capturer = handle;
    }

    handle(error) {
      return this.capturer(error);
    }

  }

  class Recover extends Catch {
    constructor(task, regain) {
      super(task);
      this.regain = regain;
    }
    handle(error) {
      return new Succeed(this.regain(error));
    }

  }

  class Format extends Catch {
    constructor(task, formatter) {
      super(task);
      this.formatter = formatter;
    }
    handle(error) {
      return new Fail(this.formatter(error));
    }

  }

  const noop = () => void 0;

  let nextID = 0;

  class Thread {
    static fork(task, onSucceed, onFail) {
      const process = new Thread();
      process.id = ++nextID;
      process.position = 0;
      process.root = task;
      process.stack = [];
      process.mailbox = [];
      process.abortHandle = null;
      process.isActive = true;
      process.isPending = false;
      process.isPaused = true;
      process.success = null;
      process.failure = null;
      process.succeed = onSucceed;
      process.fail = onFail;
      process.onSucceed = process.onSucceed.bind(process);
      process.onFail = process.onFail.bind(process);
      process.schedule();
      return process;
    }
    onSucceed(ok) {
      if (this.isPending) {
        this.isPending = false;
        this.abortHandle = null;

        if (this.success != null) {
          this.success.value = ok;
        } else {
          this.success = new Succeed(ok);
        }

        this.root = this.success;
        this.schedule();
      }
    }
    onFail(failure) {
      if (this.isPending) {
        this.isPending = false;
        this.abortHandle = null;

        if (this.failure != null) {
          this.failure.error = failure;
        } else {
          this.failure = new Fail(failure);
        }

        this.root = this.failure;
        this.schedule();
      }
    }
    kill(exit) {
      if (this.isActive) {
        this.isActive = false;
        if (this.root.abort) {
          this.root.abort(this.abortHandle);
        }
      }
    }
    schedule() {
      if (this.isPaused) {
        this.isPaused = false;
        this.step();
      }
    }
    step() {
      const process = this;
      while (process.isActive) {
        const root = process.root;
        switch (root.type) {
          case "Succeed":
            {
              const task = root;
              // If task succeeded skip all the error handling.
              while (process.position < process.stack.length && process.stack[process.position] instanceof Catch) {
                process.position++;
              }

              // If end of the stack is reached then break
              if (process.position >= process.stack.length) {
                if (process.succeed != null) {
                  process.succeed(task.value);
                }
                return;
              }

              // Otherwise step into next task.
              const then = process.stack[process.position++];
              if (then instanceof Then) {
                process.root = then.next(task.value);
              }

              break;
            }
          case "Fail":
            {
              const task = root;
              // If task fails skip all the chaining.
              while (process.position < process.stack.length && process.stack[process.position] instanceof Then) {
                process.position++;
              }

              // If end of the stack is reached then break.
              if (this.position >= process.stack.length) {
                if (process.fail != null) {
                  process.fail(task.error);
                }

                return;
              }

              // Otherwise step into next task.
              const _catch = process.stack[process.position++];
              if (_catch instanceof Catch) {
                process.root = _catch.handle(task.error);
              }

              break;
            }
          case "Then":
            {
              const task = root;
              if (process.position === 0) {
                process.stack.unshift(task);
              } else {
                process.stack[--process.position] = task;
              }

              process.root = task.task;

              break;
            }
          case "Catch":
            {
              const task = root;
              if (process.position === 0) {
                process.stack.unshift(task);
              } else {
                process.stack[--process.position] = task;
              }

              process.root = task.task;

              break;
            }
          default:
            {
              const task = root;
              process.isPending = true;
              process.abortHandle = task.fork(process.onSucceed, process.onFail);
              process.isPaused = process.isPending;
              if (this.isPending) {
                return;
              }

              break;
            }
        }
      }
    }
  }
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["./test", "../"], factory);
  } else if (typeof exports !== "undefined") {
    factory(require("./test"), require("../"));
  } else {
    var mod = {
      exports: {}
    };
    factory(global.test, global._);
    global.testApi = mod.exports;
  }
})(this, function (_test, _) {
  "use strict";

  var _test2 = _interopRequireDefault(_test);

  var Reflex = _interopRequireWildcard(_);

  function _interopRequireWildcard(obj) {
    if (obj && obj.__esModule) {
      return obj;
    } else {
      var newObj = {};

      if (obj != null) {
        for (var key in obj) {
          if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key];
        }
      }

      newObj.default = obj;
      return newObj;
    }
  }

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  (0, _test2.default)("test exported api", assert => {
    assert.ok(typeof Reflex.node, "function");

    assert.ok(typeof Reflex.html, "object");
    assert.ok(typeof Reflex.html.div, "function");

    assert.ok(typeof Reflex.thunk, "function");
    assert.ok(typeof Reflex.send, "function");
    assert.ok(typeof Reflex.forward, "function");

    assert.ok(typeof Reflex.Application, "function");

    assert.ok(typeof Reflex.Task.succeed, "function");
    assert.ok(typeof Reflex.Task.fail, "function");
    assert.ok(typeof Reflex.Task.io, "function");
    assert.ok(typeof Reflex.Task.onSuccess, "function");
    assert.ok(typeof Reflex.Task.onFailure, "function");
    assert.ok(typeof Reflex.Task.perform, "function");
    assert.ok(typeof Reflex.Task.run, "function");
  });
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "tape"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("tape"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.tape);
    global.test = mod.exports;
  }
})(this, function (exports, _tape) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var tape = _interopRequireWildcard(_tape);

  function _interopRequireWildcard(obj) {
    if (obj && obj.__esModule) {
      return obj;
    } else {
      var newObj = {};

      if (obj != null) {
        for (var key in obj) {
          if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key];
        }
      }

      newObj.default = obj;
      return newObj;
    }
  }

  exports.default = (description, unit) => tape.test(description, test => {
    var result = unit(test);
    if (result && result.then) {
      result.then(_ => test.end(), error => test.end(error || true));
    } else {
      test.end();
    }
  });
});
