// reflex/dom.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.thunk = exports.elementNS = exports.element = exports.text = exports.LazyRoot = undefined;

var _reflexDriver = require("reflex-driver");

class LazyRoot {
  constructor(view, state, mailbox) {
    this.state = state;
    this.view = view;
    this.mailbox = mailbox;
  }
  renderWith(renderer) {
    driver = renderer;
    return this.view(this.state, this.mailbox);
  }
}

exports.LazyRoot = LazyRoot;
class ErrorDriver {
  createElement(..._) {
    throw new Error(`You need to use a reflex driver to create element nodes`);
  }
  createElementNS(..._) {
    throw new Error(`You need to use a reflex driver to create element nodes`);
  }
  createTextNode(..._) {
    throw new Error(`You need to use a reflex driver to create text nodes`);
  }
  createThunk(..._) {
    throw new Error(`You need to use a reflex driver to create thunk nodes`);
  }
  render(node) {
    throw new Error(`You need to use a reflex driver to render nodes`);
  }
}

let driver = new ErrorDriver();

const text = exports.text = content => driver.createTextNode(content);

const element = exports.element = (tagName, properties, children) => driver.createElement(tagName, properties, children);

const elementNS = exports.elementNS = (namespaceURI, tagName, properties, children) => driver.createElementNS(namespaceURI, tagName, properties, children);

const thunk = exports.thunk = (key, view, ...args) => driver.createThunk(key, view, args);
