// reflex/effects.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Effects = undefined;

var _task = require("./task");

const raise = error => {
  throw Error(`Effects should be created from task that empty fail but it did fail with error ${error}`);
};

const ignore = _ => void 0;

const nil = _task.Task.succeed(void 0);

const empty = new _task.Task((succeed, fail) => void 0);

class Effects {
  static task(task) {
    console.warn("Effects.task is deprecated please use Effects.perform instead");
    return new Perform(task);
  }
  static perform(task) {
    return new Perform(task);
  }
  static tick(tag) {
    console.warn("Effects.tick is deprecated please use Effects.perform(Task.requestAnimationFrame().map(tag)) instead");
    return new Perform(_task.Task.requestAnimationFrame().map(tag));
  }
  static receive(action) {
    const fx = new Perform(new _task.Task((succeed, fail) => void Promise.resolve(action).then(succeed, fail)));
    return fx;
  }
  static batch(effects) {
    return new Batch(effects);
  }
  map(f) {
    throw Error("Subclass of abstract Effect must implement map");
  }
  execute(address) {
    throw Error("Subclass of abstract Effect must implement execute");
  }
}

exports.Effects = Effects;
class Perform extends Effects {
  constructor(task) {
    super();
    this.task = task;
  }
  map(f) {
    return new Perform(this.task.map(f));
  }
  execute(address) {
    return this.task.chain(value => _task.Task.send(address, value));
  }
}

class None extends Effects {
  map(f) {
    return Effects.none;
  }
  execute(address) {
    return nil;
  }
}
Effects.none = new None();

class Batch extends Effects {
  constructor(effects) {
    super();
    this.effects = effects;
  }
  map(f) {
    return new Batch(this.effects.map(effect => effect.map(f)));
  }
  execute(address) {
    return new _task.Task((succeed, fail) => {
      const { effects } = this;
      const count = effects.length;
      let index = 0;
      while (index < count) {
        const effect = effects[index];
        if (!(effect instanceof None)) {
          _task.Task.fork(effect.execute(address), ignore, raise);
        }

        index = index + 1;
      }
      succeed(void 0);
    });
  }

}
