// reflex/subscription.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.unsubscribe = exports.subscribe = exports.Subscribe = exports.Subscription = undefined;

var _effects = require("./effects");

class Subscription {
  static batch(subscriptions) {
    return new Batch(subscriptions);
  }
  map(tag) {
    return this.map(tag);
  }
  reduce(reducer, init) {
    return init;
  }
}

exports.Subscription = Subscription;
class Subscribe extends Subscription {

  constructor(feed, detail, tagger) {
    super();
    this.feed = feed;
    this.detail = detail;
    this.tagger = tagger;
  }
  map(tag) {
    const decoder = value => tag(this.tagger(value));
    return new Subscribe(this.feed, this.detail, decoder);
  }
  reduce(reducer, init) {
    return reducer(init, this);
  }
}

exports.Subscribe = Subscribe;
class Batch extends Subscription {
  constructor(subscriptions) {
    super();
    this.subscriptions = subscriptions;
  }
  map(tag) {
    const subscriptions = this.subscriptions.map($ => $.map(tag));
    return new Batch(subscriptions);
  }
  reduce(reducer, init) {
    return this.subscriptions.reduce((result, subscription) => subscription.reduce(reducer, result), init);
  }
}

const subscribe = exports.subscribe = (feed, detail, tagger) => new Subscribe(feed, detail, tagger);

const none = new Batch([]);
const unsubscribe = exports.unsubscribe = _ => none;
Subscription.none = none;
