// reflex/task.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Task = undefined;

var _preemptiveAnimationFrame = require("./preemptive-animation-frame");

const raise = error => {
  throw Error(`Task was not supposet to never fail but it did fail with error ${error}`);
};

const ignore = _ => void 0;

class Task {
  static create(execute) {
    console.warn("Task.create is deprecated API use new Task instead");
    return new Task(execute);
  }
  static future(request) {
    console.warn("Task.future is deprecated API use new Task instead");
    return new Future(request);
  }
  static succeed(value) {
    return new Succeed(value);
  }

  static fail(error) {
    return new Fail(error);
  }

  static spawn(task) {
    return new Spawn(task);
  }

  static sleep(time) {
    return new Sleep(time);
  }

  static requestAnimationFrame() {
    return new AnimationFrame();
  }

  static send(address, message) {
    return new Send(address, message);
  }

  static fork(task, onSucceed, onFail) {
    return Thread.fork(task, onSucceed, onFail);
  }

  static perform(task) {
    Thread.fork(task, ignore, raise);
  }

  constructor(execute, cancel) {
    this.type = "Task";
    const task = this;
    if (execute != null) {
      task.fork = execute;
    }
    if (cancel != null) {
      task.abort = cancel;
    }
  }
  chain(next) {
    return new Chain(this, next);
  }
  map(f) {
    return new Map(this, f);
  }
  capture(handle) {
    return new Capture(this, handle);
  }
  format(f) {
    return new Format(this, f);
  }
  recover(regain) {
    return new Recover(this, regain);
  }
  fork(succeed, fail) {
    return this.execute(succeed, fail);
  }
  abort(token) {
    return this.cancel(token);
  }

}

exports.Task = Task;
class Succeed extends Task {
  constructor(value) {
    super();
    this.type = "Succeed";
    this.value = value;
  }
  fork(succeed, fail) {
    succeed(this.value);
  }

}

class Fail extends Task {
  constructor(error) {
    super();
    this.type = "Fail";
    this.error = error;
  }
  fork(succeed, fail) {
    fail(this.error);
  }

}

class Sleep extends Task {
  constructor(time) {
    super();
    this.time = time;
  }
  fork(succeed, fail) {
    return setTimeout(succeed, this.time, void 0);
  }
  abort(id) {
    clearTimeout(id);
  }

}

class AnimationFrame extends Task {
  constructor() {
    super();
  }
  fork(succeed, fail) {
    return (0, _preemptiveAnimationFrame.requestAnimationFrame)(succeed);
  }
  abort(id) {
    (0, _preemptiveAnimationFrame.cancelAnimationFrame)(id);
  }
}

let threadID = 0;
class Spawn extends Task {
  constructor(task) {
    super();
    this.task = task;
  }
  fork(succeed, fail) {
    Promise.resolve(null).then(_ => Task.fork(this.task, noop, noop));

    succeed(++threadID);
  }

}

class Send extends Task {
  constructor(address, message) {
    super();
    this.message = message;
    this.address = address;
  }
  fork(succeed, fail) {
    succeed(void this.address(this.message));
  }

}

class Future extends Task {
  constructor(request) {
    super();
    this.request = request;
  }
  fork(succeed, fail) {
    this.request().then(succeed, fail);
  }

}

class Then extends Task {
  constructor(task) {
    super();
    this.type = "Then";
    this.task = task;
  }
  fork(succeed, fail) {
    this.task.fork(value => void this.next(value).fork(succeed, fail), fail);
  }
  next(input) {
    throw Error("Subclass of absract Then must implement next method");
  }

}

class Chain extends Then {
  constructor(task, next) {
    super(task);
    this.chainer = next;
  }
  next(input) {
    return this.chainer(input);
  }

}

class Map extends Then {
  constructor(task, mapper) {
    super(task);
    this.mapper = mapper;
  }
  next(input) {
    return new Succeed(this.mapper(input));
  }

}

class Catch extends Task {
  constructor(task) {
    super();
    this.type = "Catch";
    this.task = task;
  }
  fork(succeed, fail) {
    this.task.fork(succeed, error => void this.handle(error).fork(succeed, fail));
  }
  handle(error) {
    throw Error("Subclass of absract Catch must implement handle method");
  }

}

class Capture extends Catch {
  constructor(task, handle) {
    super(task);
    this.capturer = handle;
  }

  handle(error) {
    return this.capturer(error);
  }

}

class Recover extends Catch {
  constructor(task, regain) {
    super(task);
    this.regain = regain;
  }
  handle(error) {
    return new Succeed(this.regain(error));
  }

}

class Format extends Catch {
  constructor(task, formatter) {
    super(task);
    this.formatter = formatter;
  }
  handle(error) {
    return new Fail(this.formatter(error));
  }

}

const noop = () => void 0;

let nextID = 0;

class Thread {
  static fork(task, onSucceed, onFail) {
    const process = new Thread();
    process.id = ++nextID;
    process.position = 0;
    process.root = task;
    process.stack = [];
    process.mailbox = [];
    process.abortHandle = null;
    process.isActive = true;
    process.isPending = false;
    process.isPaused = true;
    process.success = null;
    process.failure = null;
    process.succeed = onSucceed;
    process.fail = onFail;
    process.onSucceed = process.onSucceed.bind(process);
    process.onFail = process.onFail.bind(process);
    process.schedule();
    return process;
  }
  onSucceed(ok) {
    if (this.isPending) {
      this.isPending = false;
      this.abortHandle = null;

      if (this.success != null) {
        this.success.value = ok;
      } else {
        this.success = new Succeed(ok);
      }

      this.root = this.success;
      this.schedule();
    }
  }
  onFail(failure) {
    if (this.isPending) {
      this.isPending = false;
      this.abortHandle = null;

      if (this.failure != null) {
        this.failure.error = failure;
      } else {
        this.failure = new Fail(failure);
      }

      this.root = this.failure;
      this.schedule();
    }
  }
  kill(exit) {
    if (this.isActive) {
      this.isActive = false;
      if (this.root.abort) {
        this.root.abort(this.abortHandle);
      }
    }
  }
  schedule() {
    if (this.isPaused) {
      this.isPaused = false;
      this.step();
    }
  }
  step() {
    const process = this;
    while (process.isActive) {
      const root = process.root;
      switch (root.type) {
        case "Succeed":
          {
            const task = root;

            while (process.position < process.stack.length && process.stack[process.position] instanceof Catch) {
              process.position++;
            }

            if (process.position >= process.stack.length) {
              if (process.succeed != null) {
                process.succeed(task.value);
              }
              return;
            }

            const then = process.stack[process.position++];
            if (then instanceof Then) {
              process.root = then.next(task.value);
            }

            break;
          }
        case "Fail":
          {
            const task = root;

            while (process.position < process.stack.length && process.stack[process.position] instanceof Then) {
              process.position++;
            }

            if (this.position >= process.stack.length) {
              if (process.fail != null) {
                process.fail(task.error);
              }

              return;
            }

            const _catch = process.stack[process.position++];
            if (_catch instanceof Catch) {
              process.root = _catch.handle(task.error);
            }

            break;
          }
        case "Then":
          {
            const task = root;
            if (process.position === 0) {
              process.stack.unshift(task);
            } else {
              process.stack[--process.position] = task;
            }

            process.root = task.task;

            break;
          }
        case "Catch":
          {
            const task = root;
            if (process.position === 0) {
              process.stack.unshift(task);
            } else {
              process.stack[--process.position] = task;
            }

            process.root = task.task;

            break;
          }
        default:
          {
            const task = root;
            process.isPending = true;
            process.abortHandle = task.fork(process.onSucceed, process.onFail);
            process.isPaused = process.isPending;
            if (this.isPending) {
              return;
            }

            break;
          }
      }
    }
  }
}
